File: leetcode-js/133-clone-graph.js
/**
 * Definition for undirected graph.
 * function UndirectedGraphNode(label) {
 *     this.label = label;
 *     this.neighbors = [];   // Array of UndirectedGraphNode
 * }
 */

/**
 * @param {UndirectedGraphNode} graph
 * @return {UndirectedGraphNode}
 */
var cloneGraph = function(graph) {
    'use strict';

    if (!graph) {
        return null;
    }

    const newHead = new UndirectedGraphNode(graph.label);
    const map = new Map();
    map.set(graph, newHead);
    const queue = [graph];

    while (queue.length) {
        const node = queue.pop();
        const clonedNode = map.get(node);

        for (let i = 0; i < node.neighbors.length; i += 1) {
            const neighbor = node.neighbors[i];
            let clonedNeighbor = map.get(neighbor);
            if (!clonedNeighbor) {
                clonedNeighbor = new UndirectedGraphNode(neighbor.label);
                map.set(neighbor, clonedNeighbor);
                queue.push(neighbor);
            }
            clonedNode.neighbors.push(clonedNeighbor);
        }
    }

    return newHead;
};

File: leetcode-java/133-clone-graph.java
/**
 * Definition for undirected graph.
 * class UndirectedGraphNode {
 *     int label;
 *     List<UndirectedGraphNode> neighbors;
 *     UndirectedGraphNode(int x) { label = x; neighbors = new ArrayList<UndirectedGraphNode>(); }
 * };
 */
public class Solution {
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        Map<Integer, UndirectedGraphNode> map = new HashMap<>();

        return dfs(map, node);
    }

    UndirectedGraphNode dfs(Map<Integer, UndirectedGraphNode> map, UndirectedGraphNode node) {
        if (node == null) {
            return null;
        }

        if (map.get(node.label) != null) {
            return map.get(node.label);
        }

        UndirectedGraphNode clone = new UndirectedGraphNode(node.label);

        map.put(node.label, clone);

        for (UndirectedGraphNode neighbor : node.neighbors) {
            clone.neighbors.add(dfs(map, neighbor));
        }

        return clone;
    }
}
